R/estimateDiversity.R

```diff
@@ -239,17 +294,70 @@ setMethod("estimateDiversity", signature = c(x = "SummarizedExperiment"),
     vegan::fisher.alpha(t(mat))
 }
 
+.calc_pd <- function(mat, tree, ...){
+
+    # Gets name of the taxa
+    taxa <- rownames(mat)
+    # Gets name of the samples
+    samples <- colnames(mat)
+
+    # Loops through every sample
+    PDs <- BiocParallel::bplapply(samples, function(sample){
+
+        # Stores taxa that are present in the sample, i.e., abundance is over 0
+        present <- taxa[ mat[, sample] > 0 ]
+        # Stores taxa that are not present, i.e., those taxa that are not in present taxa
+        absent <- taxa[ which( !(taxa %in% present) ) ]
+
+        # If no taxa are preent, then PD = 0
+        if( length(present) == 0 ){
+            PD <- 0
+        }
+
+        # If only one taxon is present
+        else if( length(present) == 1 ){
+            # Then PD is the age node of the taxon
+            PD <- tree$ages[which(tree$edge[, 2] == which(tree$tip.label == present))]
+            # If tree does not include age information, assign NA
+            if( is.null(PD) ){
+                PD <- NA
+            }
+        }
+
+        # If all taxa are present in the sample
+        else if( length(absent) == 0 ){
+            # Then PD is the sum of lengths of edges
+            PD <- sum(tree$edge.length)
+        }
+
+        # If there are taxa that are not present
+        else{
+            # Absent taxa are dropped from the tree
+            sub_tree <- ape::drop.tip(tree, absent)
+            # PD is now calculated based on the sub tree
+            PD <- sum(sub_tree$edge.length)
+        }
+    })
```

Comment:
I think there is a chance that you can avoid the `lapply` and work with vectorised subsetting. Maybe like this (untested):

```r
PD <- rep(NA,length(samples))
PD[length(present) == 0] <- 0
f <- length(present) == 1
# the match has to be reworked, since the length of the elements probably differ. Maybe this need to put in a `lapply`
PD[f] <- tree$ages[which(tree$edge[, 2] == which(tree$tip.label == present[f]))] 
PD[length(absent) == 0] <- sum(tree$edge.length)
f <- length(absent) == 1
trees <- lapply(absent[f], ape::drop.tip, why = tree)
PD[f] <- vapply(trees, function(t){sum(t$edge_length)},numeric(1))
```

Reply:
Implemented this